cmd/common/frontends/console/console.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package console
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/term"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/frontends"
+	commonutils "github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+)
+
+type frontend struct {
+	ctx        context.Context
+	cancel     func()
+	log        logger.Logger
+	isTerminal bool
+}
+
+func NewFrontend() frontends.Frontend {
+	ctx, cancel := context.WithCancel(context.Background())
+	stop := make(chan os.Signal, 1)
+	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-stop
+		log.Debugf("got signal, cancelling context")
+		cancel()
+		<-stop
+		log.Debugf("got signal again, exiting")
+		// On second signal, exit forcefully
+		os.Exit(1)
```

Comment:
If another package uses the same code pattern, there would be two channels for the signals and it will be delivered to both in a non-determinist order. But then, `os.Exit()` would terminate the program immediately, so the other handler will not complete.
https://go.dev/play/p/Bfg15_DV7nA

I wonder if it's ok to use signal.Notify() in a package. Maybe NewFrontend() could accept the `ctx` as parameter and the caller would set-up the signal handlers?

Reply:
IMHO, I think that package being called `Frontend` assumes it is running at the topmost layer and handling exactly things like user input/output and also signals, reacting and informing lower layers on-demand (using the ctx).
Which other package do you think would listen for these signals in the same way (with "legitimate" interest)?

> I wonder if it's ok to use signal.Notify() in a package. Maybe NewFrontend() could accept the ctx as parameter and the caller would set-up the signal handlers?

Hmm, the console frontend is only shared between `local-gadget` and `kubectl-gadget` - but yeah, we could move the signal handling to the `registry.go` if you think it improves readability.